### PR TITLE
POI attachment changes

### PR DESCRIPTION
**Pull request by cannonfodder1**
Changes the backend so POIs are attached to the mission on spawning, not on completion
CC #325

### CODE REVIEW

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2ActivityTemplate_Mission.uc [hunk 1]
```diff
@@ -162,6 +165,7 @@ static function bool DefaultShouldProgressChain (XComGameState_Activity Activity
 defaultproperties
 {
 	bAssignFactionToMissionSite = true
+	bNeedsPOI = false
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2ActivityTemplate_Mission.uc]: This is redundant

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc [hunk 2]
```diff
@@ -1160,4 +1160,29 @@ static protected function int SortMultiStepLerpSteps (MultiStepLerpStep A, Multi
 	if (A.X == B.X) return 0;
 
 	return A.X < B.X ? 1 : -1;
-}
\ No newline at end of file
+}
+
+static function HandlePostMissionPOI(XComGameState NewGameState, XComGameState_Activity ActivityState, XComGameState_MissionSite MissionState, bool bSuccess)
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc]: No need to pass the mission state, just use `GetMissionStateFromActivity(ActivityState)`

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc [hunk 3]
```diff
@@ -1160,4 +1160,29 @@ static protected function int SortMultiStepLerpSteps (MultiStepLerpStep A, Multi
...
+			{
+				ResHQ = class'UIUtilities_Strategy'.static.GetResistanceHQ();
+				MissionState.POIToSpawn = ResHQ.ChoosePOI(NewGameState);
+				`CI_Trace(ActivityState.GetMyTemplateName() $ " has no POI, creating a replacement");
+			}
+		
+			class'X2StrategyElement_DefaultMissionSources'.static.SpawnPointOfInterest(NewGameState, MissionState);
+		}
+		
+		class'XComGameState_HeadquartersResistance'.static.DeactivatePOI(NewGameState, MissionState.POIToSpawn);
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc]: This needs to be inside an `else` branch

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc [hunk 4]
```diff
@@ -1160,4 +1160,29 @@ static protected function int SortMultiStepLerpSteps (MultiStepLerpStep A, Multi
...
+	Template = X2ActivityTemplate_Mission(ActivityState.GetMyTemplate());
+
+	if (Template != none && Template.bNeedsPOI)
+	{
+		if (bSuccess)
+		{
+			if (MissionState.POIToSpawn.ObjectID <= 0)
+			{
+				ResHQ = class'UIUtilities_Strategy'.static.GetResistanceHQ();
+				MissionState.POIToSpawn = ResHQ.ChoosePOI(NewGameState);
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc]: You can simply do `MissionState.PickPOI()`

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2ActivityTemplate_Mission.uc [hunk 5]
```diff
@@ -99,11 +100,13 @@ static function GenericOnSuccess (XComGameState NewGameState, XComGameState_Acti
 
 	MissionState = class'X2Helper_Infiltration'.static.GetMissionStateFromActivity(ActivityState);
 	class'X2StrategyElement_DefaultMissionSources'.static.GiveRewards(NewGameState, MissionState);
+	class'X2Helper_Infiltration'.static.HandlePostMissionPOI(NewGameState, ActivityState, MissionState, true);
 	MissionState.RemoveEntity(NewGameState);
 
 	ActivityState = XComGameState_Activity(NewGameState.ModifyStateObject(class'XComGameState_Activity', ActivityState.ObjectID));
 	ActivityState.MarkSuccess(NewGameState);
 	
+	class'XComGameState_HeadquartersResistance'.static.DeactivatePOI(NewGameState, MissionState.POIToSpawn);
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2ActivityTemplate_Mission.uc]: This should not be here

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2ActivityTemplate_Mission.uc [hunk 6]
```diff
@@ -112,11 +115,13 @@ static function GenericOnFailure (XComGameState NewGameState, XComGameState_Acti
 	local XComGameState_MissionSite MissionState;
 	
 	MissionState = class'X2Helper_Infiltration'.static.GetMissionStateFromActivity(ActivityState);
+	class'X2Helper_Infiltration'.static.HandlePostMissionPOI(NewGameState, ActivityState, MissionState, false);
 	MissionState.RemoveEntity(NewGameState);
 	
 	ActivityState = XComGameState_Activity(NewGameState.ModifyStateObject(class'XComGameState_Activity', ActivityState.ObjectID));
 	ActivityState.MarkFailed(NewGameState);
 	
+	class'XComGameState_HeadquartersResistance'.static.DeactivatePOI(NewGameState, MissionState.POIToSpawn);
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2ActivityTemplate_Mission.uc]: This should not be here

**Comment by cannonfodder1** [comment]
Should this PR close #325 or is there other stuff yet to do on it?

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc [hunk 7]
```diff
@@ -1155,4 +1160,36 @@ static protected function int SortMultiStepLerpSteps (MultiStepLerpStep A, Multi
...
\ No newline at end of file
+}
+
+static function HandlePostMissionPOI(XComGameState NewGameState, XComGameState_Activity ActivityState, bool bSuccess)
+{
+	local XComGameState_HeadquartersResistance ResHQ;
+	local XComGameState_MissionSite MissionState;
+	local X2ActivityTemplate_Mission Template;
+
+	MissionState = GetMissionStateFromActivity(ActivityState);
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc]: It's best to call `NewGameState.ModifyStateObject` here.

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc [hunk 8]
```diff
@@ -1155,4 +1160,36 @@ static protected function int SortMultiStepLerpSteps (MultiStepLerpStep A, Multi
...
+	Template = X2ActivityTemplate_Mission(ActivityState.GetMyTemplate());
+
+	if (Template != none && Template.bNeedsPOI)
+	{
+		if (bSuccess)
+		{
+			if (MissionState.POIToSpawn.ObjectID <= 0)
+			{
+				ResHQ = class'UIUtilities_Strategy'.static.GetResistanceHQ();
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc]: Unused variable

File: CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc [hunk 9]
```diff
@@ -1155,4 +1160,36 @@ static protected function int SortMultiStepLerpSteps (MultiStepLerpStep A, Multi
...
+				MissionState.PickPOI(NewGameState);
+				`CI_Trace(ActivityState.GetMyTemplateName() $ " has no POI, creating a replacement");
```
Comment (Xymanek) [inline CovertInfiltration/Src/CovertInfiltration/Classes/X2Helper_Infiltration.uc]: This should probably be a warning and include something like `but requires one`

**Comment by Xymanek** [comment]
I'll close #325 manually with some comments

**Pull request merged**